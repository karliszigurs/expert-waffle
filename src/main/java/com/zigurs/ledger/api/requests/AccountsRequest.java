package com.zigurs.ledger.api.requests;

import lombok.NonNull;

import java.util.UUID;

public record AccountsRequest(@NonNull UUID client_id) {
}
