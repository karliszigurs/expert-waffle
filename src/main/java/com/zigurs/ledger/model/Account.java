package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigInteger;
import java.util.UUID;

@Getter
@Entity
@Table(name = "accounts")
@AllArgsConstructor
@NoArgsConstructor
public class Account {

    @Id
    @UuidGenerator
    private UUID id;

    @ManyToOne
    private Client client;

    private String currency;

    @Setter
    private BigInteger balance;
}
