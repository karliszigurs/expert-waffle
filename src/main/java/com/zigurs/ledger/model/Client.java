package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.UuidGenerator;

import java.util.UUID;

@Entity
@Table(name = "clients")
@AllArgsConstructor
@NoArgsConstructor
public class Client {

    @Id
    @UuidGenerator
    @Getter
    private UUID id;
}