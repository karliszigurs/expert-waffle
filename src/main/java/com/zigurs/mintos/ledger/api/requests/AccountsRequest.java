package com.zigurs.mintos.ledger.api.requests;

import lombok.NonNull;

import java.util.UUID;

public record AccountsRequest(@NonNull UUID client_id) {
}
