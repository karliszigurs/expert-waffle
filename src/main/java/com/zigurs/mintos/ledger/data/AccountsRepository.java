package com.zigurs.mintos.ledger.data;

import com.zigurs.mintos.ledger.model.Account;

import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface AccountsRepository extends Repository<Account, UUID> {

    List<Account> findByClientId(UUID client_id);

    Optional<Account> findById(UUID uuid);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select a from Account a where a.id = :id")
    Optional<Account> findByIdForUpdate(UUID id);

    Account save(Account account);
}
