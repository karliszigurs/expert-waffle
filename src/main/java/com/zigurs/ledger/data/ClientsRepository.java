package com.zigurs.ledger.data;

import com.zigurs.ledger.model.Client;

import org.springframework.data.repository.Repository;

import java.util.Optional;
import java.util.UUID;

public interface ClientsRepository extends Repository<Client, UUID> {

    Optional<Client> findById(UUID id);
}
