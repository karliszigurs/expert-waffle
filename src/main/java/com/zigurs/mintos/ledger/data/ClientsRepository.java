package com.zigurs.mintos.ledger.data;

import com.zigurs.mintos.ledger.model.Client;

import org.springframework.data.repository.Repository;

import java.util.Optional;
import java.util.UUID;

public interface ClientsRepository extends Repository<Client, UUID> {

    Optional<Client> findById(UUID id);

}
