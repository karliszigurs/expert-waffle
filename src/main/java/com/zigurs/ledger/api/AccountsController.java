package com.zigurs.ledger.api;

import com.zigurs.ledger.api.exceptions.NotFoundException;
import com.zigurs.ledger.api.requests.AccountsRequest;
import com.zigurs.ledger.api.responses.AccountView;
import com.zigurs.ledger.data.AccountsRepository;
import com.zigurs.ledger.data.ClientsRepository;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

/**
 * Accounts controller following Spring Boot patterns.
 */
@RestController
public class AccountsController {

    private final ClientsRepository clientsRepository;
    private final AccountsRepository accountsRepository;

    public AccountsController(
            ClientsRepository clientsRepository,
            AccountsRepository accountsRepository
    ) {
        this.clientsRepository = clientsRepository;
        this.accountsRepository = accountsRepository;
    }

    @PostMapping("accounts")
    public ResponseEntity<List<AccountView>> getAccounts(@RequestBody AccountsRequest request) {
        // Validate client id to differentiate between invalid client vs client, but without accounts.
        clientsRepository.findById(request.client_id()).orElseThrow(
                () -> new NotFoundException("client not found")
        );

        // We know the client id is valid here, we are guaranteed to get at least an empty list
        return ResponseEntity.ok(
                accountsRepository
                        .findByClientId(request.client_id())
                        .stream()
                        .map(AccountView::fromModel)
                        .toList()
        );
    }
}
