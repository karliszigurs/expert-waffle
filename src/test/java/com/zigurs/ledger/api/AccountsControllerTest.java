package com.zigurs.ledger.api;

import com.zigurs.ledger.api.requests.AccountsRequest;
import com.zigurs.ledger.data.AccountsRepository;
import com.zigurs.ledger.data.ClientsRepository;
import com.zigurs.ledger.model.Account;
import com.zigurs.ledger.model.Client;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import testutils.Snapshot;

import java.math.BigInteger;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static io.github.jsonSnapshot.SnapshotMatcher.*;
import static org.hamcrest.core.StringContains.containsString;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static testutils.Utils.asJsonString;

@WebMvcTest(AccountsController.class)
class AccountsControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private ClientsRepository clientsRepository;

    @MockBean
    private AccountsRepository accountsRepository;

    @BeforeAll
    public static void beforeAll() {
        start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    void getAccountsHappyPath() throws Exception {
        Client client = new Client(UUID.randomUUID());
        Account account = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(clientsRepository.findById(client.getId())).thenReturn(
                Optional.of(client)
        );

        when(accountsRepository.findByClientId(client.getId())).thenReturn(
                List.of(account)
        );

        AccountsRequest req = new AccountsRequest(client.getId());

        MvcResult result = mockMvc.perform(
                        post("/accounts")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isOk())
                .andExpect(content().string(containsString(account.getId().toString())))
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void getAccountsNoAccounts() throws Exception {
        Client client = new Client(UUID.randomUUID());

        when(clientsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(client)
        );

        when(accountsRepository.findByClientId(any(UUID.class))).thenReturn(
                List.of()
        );

        AccountsRequest req = new AccountsRequest(UUID.randomUUID());

        MvcResult result = mockMvc.perform(
                        post("/accounts")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isOk())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void getAccountsInvalidClient() throws Exception {
        when(clientsRepository.findById(any(UUID.class))).thenReturn(
                Optional.empty()
        );

        AccountsRequest req = new AccountsRequest(UUID.randomUUID());

        MvcResult result = mockMvc.perform(
                        post("/accounts")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isNotFound())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void getAccountsErrorWhenFetchingClient() throws Exception {
        when(clientsRepository.findById(any(UUID.class))).thenThrow(
                new RuntimeException()
        );

        AccountsRequest req = new AccountsRequest(UUID.randomUUID());

        MvcResult result = mockMvc.perform(
                        post("/accounts")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is5xxServerError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void getAccountsErrorWhenFetchingAccounts() throws Exception {
        Client client = new Client(UUID.randomUUID());

        when(clientsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(client)
        );

        when(accountsRepository.findByClientId(any(UUID.class))).thenThrow(
                new RuntimeException()
        );

        AccountsRequest req = new AccountsRequest(UUID.randomUUID());

        MvcResult result = mockMvc.perform(
                        post("/accounts")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is5xxServerError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }
}