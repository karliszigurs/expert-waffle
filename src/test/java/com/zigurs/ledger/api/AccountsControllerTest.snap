com.zigurs.ledger.api.AccountsControllerTest.getAccountsErrorWhenFetchingAccounts=[
  "{\"error\":\"internal error\"}"
]


com.zigurs.ledger.api.AccountsControllerTest.getAccountsErrorWhenFetchingClient=[
  "{\"error\":\"internal error\"}"
]


com.zigurs.ledger.api.AccountsControllerTest.getAccountsHappyPath=[
  "[{\"account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"currency\":\"EUR\",\"balance\":1}]"
]


com.zigurs.ledger.api.AccountsControllerTest.getAccountsInvalidClient=[
  "{\"error\":\"404 client not found\"}"
]


com.zigurs.ledger.api.AccountsControllerTest.getAccountsNoAccounts=[
  "[]"
]